File: 2025/days/day09.py
from pathlib import Path
import time

HORIZ = 1
VERT = 2


def main():
    tiles = [
        [int(num) for num in line.split(',')]
        for line in Path("inputs/day09.txt").read_text().splitlines()
    ]
    lines = [(tiles[i], tiles[i+1]) for i in range(len(tiles) - 1)]
    lines.append((tiles[0], tiles[-1]))

    max_size = 0
    max_green_size = 0
    for i in range(0, len(tiles) - 1):
        for j in range(i + 1, len(tiles)):
            a = tiles[i]
            b = tiles[j]
            size = (abs(a[0] - b[0]) + 1) * (abs(a[1] - b[1]) + 1)
            if size > max_size:
                max_size = size
            if is_green(a, b, lines) and size > max_green_size:
                max_green_size = size
    print(f"Part 1: {max_size}")
    print(f"Part 2: {max_green_size}")


def is_green(a, b, lines) -> bool:
    min_x = min(a[0], b[0])
    max_x = max(a[0], b[0])
    min_y = min(a[1], b[1])
    max_y = max(a[1], b[1])

    top_left = (min_x, min_y)
    bottom_right = (max_x, max_y)
    tl_int: dict | None = {'left': None,
                           'right': None, 'up': None, 'down': None}
    br_int: dict | None = {'left': None,
                           'right': None, 'up': None, 'down': None}

    for line in lines:
        # VERTICALS
        if get_line_type(line) == VERT:
            # top left
            if between(top_left[1], line[0][1], line[1][1]):
                intersection = (line[0][0], top_left[1])
                on_line_end = intersection[1] == line[0][1] or intersection[1] == line[1][1]
                if intersection[0] <= top_left[0]:
                    # left
                    tl_int['left'] = {
                        'intersection': intersection, 'inside': not on_line_end}
                if intersection[0] > top_left[0]:
                    # right
                    if tl_int['right'] == None or tl_int['right']['intersection'] == top_left:
                        tl_int['right'] = {
                            'intersection': intersection, 'inside': not on_line_end}
                    else:
                        if on_line_end:
                            if (not tl_int['right']['inside']) and tl_int['right']['intersection'][0] < intersection[0]:
                                tl_int['right'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
                        else:
                            if (not tl_int['right']['inside']) or tl_int['right']['intersection'][0] > intersection[0]:
                                tl_int['right'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
            # bottom right
            if between(bottom_right[1], line[0][1], line[1][1]):
                intersection = (line[0][0], bottom_right[1])
                on_line_end = intersection[1] == line[0][1] or intersection[1] == line[1][1]
                if intersection[0] < bottom_right[0]:
                    # left
                    if br_int['left'] == None or br_int['left']['intersection'] == bottom_right:
                        br_int['left'] = {
                            'intersection': intersection, 'inside': not on_line_end}
                    else:
                        if on_line_end:
                            if (not br_int['left']['inside']) and br_int['left']['intersection'][0] > intersection[0]:
                                br_int['left'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
                        else:
                            if (not br_int['left']['inside']) or br_int['left']['intersection'][0] < intersection[0]:
                                br_int['left'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
                if intersection[0] >= bottom_right[0]:
                    # right
                    br_int['right'] = {
                        'intersection': intersection, 'inside': not on_line_end}
        # HORIZONTALS
        else:
            # top left
            if between(top_left[0], line[0][0], line[1][0]):
                intersection = (top_left[0], line[0][1])
                on_line_end = intersection[0] == line[0][0] or intersection[0] == line[1][0]
                if intersection[1] <= top_left[1]:
                    # up
                    tl_int['up'] = {'intersection': intersection,
                                    'inside': not on_line_end}
                if intersection[1] > top_left[1]:
                    # down
                    if tl_int['down'] == None or tl_int['down']['intersection'] == top_left:
                        tl_int['down'] = {
                            'intersection': intersection, 'inside': not on_line_end}
                    else:
                        if on_line_end:
                            if (not tl_int['down']['inside']) and tl_int['down']['intersection'][1] < intersection[1]:
                                tl_int['down'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
                        else:
                            if (not tl_int['down']['inside']) or tl_int['down']['intersection'][1] > intersection[1]:
                                tl_int['down'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
            # bottom right
            if between(bottom_right[0], line[0][0], line[1][0]):
                intersection = (bottom_right[0], line[0][1])
                on_line_end = intersection[0] == line[0][0] or intersection[0] == line[1][0]
                if intersection[1] < bottom_right[1]:
                    # up
                    if br_int['up'] == None or br_int['up']['intersection'] == bottom_right:
                        br_int['up'] = {
                            'intersection': intersection, 'inside': not on_line_end}
                    else:
                        if on_line_end:
                            if (not br_int['up']['inside']) and br_int['up']['intersection'][0] > intersection[0]:
                                br_int['up'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
                        else:
                            if (not br_int['up']['inside']) or br_int['up']['intersection'][0] < intersection[0]:
                                br_int['up'] = {
                                    'intersection': intersection, 'inside': not on_line_end}
                if intersection[1] >= bottom_right[1]:
                    # down
                    br_int['down'] = {
                        'intersection': intersection, 'inside': not on_line_end}

    if any(val == None for val in tl_int.values()) or any(val == None for val in br_int.values()):
        return False
    
    if (
        tl_int['right']['intersection'][0] < max_x or
        tl_int['down']['intersection'][1] < max_y or
        br_int['left']['intersection'][0] > min_x or
        br_int['up']['intersection'][1] > min_y
    ):
        return False
    
    return True


def get_line_type(line):
    if line[0][1] == line[1][1]:
        return HORIZ
    return VERT


def between(val, a, b) -> bool:
    start = min(a, b)
    end = max(a, b)
    return val >= start and val <= end


if __name__ == "__main__":
    start = time.time()
    main()
    end = time.time()
    elapsed = end - start
    print(f"Execution time: {elapsed:.4f} seconds")

File: 2025/days/day09.js
const fs = require("fs");

console.time();

const HORIZ = 1;
const VERT = 2;

let tiles = fs.readFileSync("inputs/day09.txt", "utf-8")
  .split("\n")
  .map((line) => line.split(",").map((num) => parseInt(num)));
let lines = [];
for (let i = 0; i < tiles.length - 1; i++) lines.push([tiles[i], tiles[i + 1]]);
lines.push([tiles[0], tiles.at(-1)]);

let maxSize = 0;
let maxGreenSize = 0;
for (let i = 0; i < tiles.length - 1; i++) {
  for (let j = i + 1; j < tiles.length; j++) {
    let a = tiles.at(i);
    let b = tiles[j];
    let size = (Math.abs(a[0] - b[0]) + 1) * (Math.abs(a[1] - b[1]) + 1);
    if (size > maxSize) maxSize = size;
    if (isGreen(a, b, lines) && size > maxGreenSize) maxGreenSize = size;
  }
}
console.log(`Part 1: ${maxSize}`);
console.log(`Part 2: ${maxGreenSize}`);

console.timeEnd();

function isGreen(a, b, lines) {
  let minX = Math.min(a[0], b[0]);
  let maxX = Math.max(a[0], b[0]);
  let minY = Math.min(a[1], b[1]);
  let maxY = Math.max(a[1], b[1]);

  let topLeft = [minX, minY];
  let bottomRight = [maxX, maxY];
  let tlInt = { left: null, right: null, up: null, down: null };
  let brInt = { left: null, right: null, up: null, down: null };

  for (let line of lines) {
    // VERTICALS
    if (getLineType(line) == VERT) {
      // top left
      if (between(topLeft[1], line[0][1], line[1][1], true)) {
        // intersection found
        let intersection = [line[0][0], topLeft[1]];
        let onLineEnd =
          intersection[1] == line[0][1] || intersection[1] == line[1][1];
        if (intersection[0] <= topLeft[0]) {
          // left
          tlInt.left = { intersection, inside: !onLineEnd };
        }
        if (intersection[0] > topLeft[0]) {
          // right
          if (
            tlInt.right === null ||
            pointsEqual(tlInt.right.intersection, topLeft)
          ) {
            tlInt.right = { intersection, inside: !onLineEnd };
          } else {
            // short or long based on inside or outside
            if (onLineEnd) {
              if (
                !tlInt.right.inside &&
                tlInt.right.intersection[0] < intersection[0]
              )
                tlInt.right = { intersection, inside: !onLineEnd };
            } else {
              if (
                !tlInt.right.inside ||
                tlInt.right.intersection[0] > intersection[0]
              )
                tlInt.right = { intersection, inside: !onLineEnd };
            }
          }
        }
      }

      // bottom right
      if (between(bottomRight[1], line[0][1], line[1][1], true)) {
        // intersection found
        let intersection = [line[0][0], bottomRight[1]];
        let onLineEnd =
          intersection[1] == line[0][1] || intersection[1] == line[1][1];
        if (intersection[0] < bottomRight[0]) {
          // left
          if (
            brInt.left === null ||
            pointsEqual(brInt.left.intersection, bottomRight)
          ) {
            brInt.left = { intersection, inside: !onLineEnd };
          } else {
            // short or long based on inside or outside
            if (onLineEnd) {
              if (
                !brInt.left.inside &&
                brInt.left.intersection[0] > intersection[0]
              )
                brInt.left = { intersection, inside: !onLineEnd };
            } else {
              if (
                !brInt.left.inside ||
                brInt.left.intersection[0] < intersection[0]
              )
                brInt.left = { intersection, inside: !onLineEnd };
            }
          }
        }
        if (intersection[0] >= bottomRight[0]) {
          // right
          brInt.right = { intersection, inside: !onLineEnd };
        }
      }

      // HORIZONTALS
    } else {
      // top left
      if (between(topLeft[0], line[0][0], line[1][0], true)) {
        // intersection found
        let intersection = [topLeft[0], line[0][1]];
        let onLineEnd =
          intersection[0] == line[0][0] || intersection[0] == line[1][0];
        if (intersection[1] <= topLeft[1]) {
          // up
          tlInt.up = { intersection, inside: !onLineEnd };
        }
        if (intersection[1] > topLeft[1]) {
          // down
          if (
            tlInt.down === null ||
            pointsEqual(tlInt.down.intersection, topLeft)
          ) {
            tlInt.down = { intersection, inside: !onLineEnd };
          } else {
            // short or long based on inside or outside
            if (onLineEnd) {
              if (
                !tlInt.down.inside &&
                tlInt.down.intersection[1] < intersection[1]
              )
                tlInt.down = { intersection, inside: !onLineEnd };
            } else {
              if (
                !tlInt.down.inside ||
                tlInt.down.intersection[1] > intersection[1]
              ) {
                tlInt.down = { intersection, inside: !onLineEnd };
              }
            }
          }
        }
      }

      // bottom right
      if (between(bottomRight[0], line[0][0], line[1][0], true)) {
        // intersection found
        let intersection = [bottomRight[0], line[0][1]];
        let onLineEnd =
          intersection[0] == line[0][0] || intersection[0] == line[1][0];
        if (intersection[1] < bottomRight[1]) {
          // up
          if (
            brInt.up === null ||
            pointsEqual(brInt.up.intersection, bottomRight)
          ) {
            brInt.up = { intersection, inside: !onLineEnd };
          } else {
            // short or long based on inside or outside
            if (onLineEnd) {
              if (
                !brInt.up.inside &&
                brInt.up.intersection[1] > intersection[1]
              )
                brInt.up = { intersection, inside: !onLineEnd };
            } else {
              if (
                !brInt.up.inside ||
                brInt.up.intersection[1] < intersection[1]
              )
                brInt.up = { intersection, inside: !onLineEnd };
            }
          }
        }
        if (intersection[1] >= bottomRight[1]) {
          // down
          brInt.down = { intersection, inside: !onLineEnd };
        }
      }
    }
  }

  // console.log("tl");
  // console.log(tlInt);
  // console.log("br");
  // console.log(brInt);

  if (
    Object.values(tlInt).some((val) => val === null) ||
    Object.values(brInt).some((val) => val === null)
  ) {
    return false;
  }

  if (
    tlInt.right.intersection[0] < maxX ||
    tlInt.down.intersection[1] < maxY ||
    brInt.left.intersection[0] > minX ||
    brInt.up.intersection[1] > minY
  )
    return false;
  return true;
}

function getLineType(line) {
  if (line[0][1] == line[1][1]) return HORIZ;
  return VERT;
}

function between(val, a, b) {
  let start = Math.min(a, b);
  let end = Math.max(a, b);
  return val >= start && val <= end;
}

function pointsEqual(a, b) {
  return a[0] === b[0] && a[1] === b[1];
}
